### PR TITLE
backend/tidb: use REPLACE INTO or INSERT IGNORE INTO to provide idempotent insertion

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
1 modifications from #242 : 
- use `REPLACE INTO` to provide idempotent insertion while PK or UNIQUE exsits

### What is changed and how it works?
replace according SQL to `REPLACE INTO` or `INSERT IGNORE INTO` statement in backend/tidb's WriteRows. 
add a config field to decide behaviour

```
[tikv-importer]
backend = "tidb"
on-duplicate = "replace" # or "ignore", "error". which represents REPLACE INTO, INSERT IGNORE INTO, INSERT INTO, respectively.
```

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression: not sure `REPLACE INTO` is slower

Related changes

### CODE REVIEW

**Comment by lance6716** [comment]
/run-all-tests

PTAL @kennytm @3pointer

**Comment by kennytm** [comment]
Could we make it configurable to choose whether to use `INSERT INTO` / `REPLACE INTO` or `INSERT IGNORE INTO`?

**Comment by lance6716** [comment]
Did a simple test using `mysql < text.sql`, `insert ignore into` is slightly faster(50% duplicated), so how about switch to `insert ignore into`?

ignore is reasonable since dumped files won't change.

**Comment by kennytm** [comment]
This would mean that, during incremental import, if the new data does conflict with old data the default behavior is the new data being discarded, which is not very intuitive.

I think the default should still be REPLACE but we could configure to INSERT or IGNORE if needed

**Comment by lance6716** [comment]
Oh I forget we should keep `INSERT INTO` as an option

**Comment by GregoryIan** [comment]
LGTM

**Comment by GregoryIan** [comment]
@kennytm PTAL

**Comment by lance6716** [comment]
> @kennytm PTAL

Kenny says we might provide 3 types of behavior: replace, insert ignore, insert. I'm agree that we might provide all 3 types, and if a further benchmark shows there's little performance difference, we might reduce to 2 or 1 type at that time.

**Comment by GregoryIan** [comment]
would lightning compute checksum many times for the rows which are executed more than twice? Rest LGTM

**Comment by kennytm** [comment]
@GregoryIan Lightning currently won't do checksum with TiDB backend (since the ADMIN CHECKSUM is based on KV pairs).

**Comment by kennytm** [comment]
@lance6716 I'll start review after you've changed the config from a bool to a string option.

**Comment by lance6716** [comment]
/run-all-tests

PTAL @kennytm

File: lightning/backend/tidb.go [hunk 1]
```diff
@@ -263,7 +265,15 @@ func (be *tidbBackend) WriteRows(ctx context.Context, _ uuid.UUID, tableName str
 	}
 
 	var insertStmt strings.Builder
-	insertStmt.WriteString("INSERT INTO ")
+	switch be.OnDuplicate {
```
Comment (GregoryIan) [inline lightning/backend/tidb.go]: what's the default behavior of  `OnDuplicate`?

Comment (GregoryIan) [reply lightning/backend/tidb.go]: should we give it a default value if users don't configure `OnDuplicate`

Comment (lance6716) [reply lightning/backend/tidb.go]: default behavior is replacing on duplicate, which specifies in `Adjust()`.

https://github.com/pingcap/tidb-lightning/pull/243/files#diff-5f8541d99c1c66dcc6e24f4418ce5c24R205

`Adjust()` will be called in `RunOnce` and `handlePostTask()` (which then goes to RunServer), cover all user's case.

Codes call `NewTiDBBackend` should specify this param manually 🤔

File: tests/tidb_duplicate_data/run.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+
```
Comment (GregoryIan) [inline tests/tidb_duplicate_data/run.sh]: In fact we need such a case more:
* lightning start to loading
* panic on some statements (inject failpoint)
* restart lightning to finish loading
* check everything is ok

File: lightning/backend/tidb.go [hunk 3]
```diff
@@ -42,14 +43,15 @@ type tidbEncoder struct {
 
 type tidbBackend struct {
 	db *sql.DB
+	OnDuplicate string
```
Comment (kennytm) [inline lightning/backend/tidb.go]: Does any code outside this module needs to access this field? If not make the field private.

```suggestion
	onDuplicate string
```

(also, `go fmt` this file).

File: lightning/backend/tidb.go [hunk 4]
```diff
@@ -42,14 +43,15 @@ type tidbEncoder struct {
...
 }
 
 // NewTiDBBackend creates a new TiDB backend using the given database.
 //
 // The backend does not take ownership of `db`. Caller should close `db`
 // manually after the backend expired.
-func NewTiDBBackend(db *sql.DB) Backend {
-	return MakeBackend(&tidbBackend{db: db})
+func NewTiDBBackend(db *sql.DB, OnDuplicate string) Backend {
```
Comment (kennytm) [inline lightning/backend/tidb.go]:
```suggestion
func NewTiDBBackend(db *sql.DB, onDuplicate string) Backend {
```

Variable names should never be in uppercase.

File: lightning/config/config.go [hunk 5]
```diff
@@ -47,6 +47,13 @@ const (
 	CheckpointDriverMySQL = "mysql"
 	// CheckpointDriverFile is a constant for choosing the "File" checkpoint driver in the configuration.
 	CheckpointDriverFile = "file"
+
+	// RepalceOnDup indicates using REPLACE INTO to insert data
+	RepalceOnDup = "replace"
```
Comment (kennytm) [inline lightning/config/config.go]:
```suggestion
	ReplaceOnDup = "replace"
```

File: lightning/backend/tidb.go (hunk 4 above)
Comment (lance6716) [reply lightning/backend/tidb.go]: my apologies, I simply delete prefix of `replaceOnDuplicate` and forget Go's semantic 😭

**Comment by lance6716** [comment]
/run-all-tests

PTAL @kennytm @GregoryIan

File: lightning/backend/tidb.go (hunk 1 above)
Comment (kennytm) [reply lightning/backend/tidb.go]: IMO the extra defense is redundant. Not a big issue anyway.

**Pull request merged**